server/cmd_auth.go

```diff
@@ -24,8 +24,21 @@ func (cmd *Select) Handle(conn Conn) error {
 		return ErrNotAuthenticated
 	}
 
+	// As per RFC1730#6.3.1,
+	// 		The SELECT command automatically deselects any
+	// 		currently selected mailbox before attempting the new selection.
+	// 		Consequently, if a mailbox is selected and a SELECT command that
+	// 		fails is attempted, no mailbox is selected.
+	// Thus if this GetMailbox call fails to find a mailbox, this is not an
+	// error and we should simply unselect.
+	// For example, some clients (e.g. Apple Mail) perform SELECT "" when the
+	// server doesn't announce the UNSELECT capability.
 	mbox, err := ctx.User.GetMailbox(cmd.Mailbox)
-	if err != nil {
+	if errors.Is(err, backend.ErrNoSuchMailbox) {
```

Comment:
Nice catch! Should we unselect on any error, or just when the mailbox isn't found?

The RFC wording makes me think the former would be better, in which case we don't need to check for `backend.ErrNoSuchMailbox`.

Reply:
I agree with emersion, there are other reasons SELECT may fail and the effect should be the same per RFC.

Reply:
Good point. I guess it's quite clear -- always unselect first, then only select if no other errors are thrown.

I don't suppose it's threadsafe enough to set Mailbox to nil at the top of the function, which would follow RFC's description even more closely.

Reply:
Hmm but should `Handle` return an error if `GetMailbox` does? Should all errors from `GetMailbox` be ignored? Not sure what's best to do here.

Reply:
Pushed changes that unselect on any SELECT failure; all errors are bubbled up.

Reply:
>Hmm but should Handle return an error if GetMailbox does? Should all errors from GetMailbox be ignored? Not sure what's best to do here.

It should be fine to always reset the mailbox, since errors returned here will be forwarded to client via an IMAP response.